crates/metal-llvm/src/codegen.rs

```diff
@@ -0,0 +1,189 @@
+    LLVMLinkage,
+};
+
+// TODO: move this to another file?
+pub struct Variable<'src> {
+    pntr: LLVMValueRef,
+    ty: metal_ast::Ty<'src>,
+}
+
+pub struct CodeGen {
+    module_name: &'static str,
+    ctx: LLVMContextRef,
+    builder: LLVMBuilderRef,
+    module: LLVMModuleRef,
+}
+
+impl CodeGen {
+    // complete codegen process
+    pub fn finish(&self) -> &'static str {
+        unsafe {
+            let ir = LLVMPrintModuleToString(self.module);
+            let cir = CStr::from_ptr(ir);
+            // TODO: specify which module specifically.
+            let string_ir = cir
+                .to_str()
+                .expect("Failed to turn LLVM module into string");
+            string_ir
+        }
+    }
+
+    pub fn primitive(&self, primitive: &metal_ast::Primitives) -> LLVMTypeRef {
+        match primitive {
+            metal_ast::Primitives::I8 => unsafe { LLVMIntTypeInContext(self.ctx, 8) },
+            metal_ast::Primitives::I16 => unsafe { LLVMIntTypeInContext(self.ctx, 16) },
+            metal_ast::Primitives::I32 => unsafe { LLVMIntTypeInContext(self.ctx, 32) },
+            metal_ast::Primitives::I64 => unsafe { LLVMIntTypeInContext(self.ctx, 64) },
+            metal_ast::Primitives::I128 => unsafe { LLVMIntTypeInContext(self.ctx, 128) },
+
+            metal_ast::Primitives::F32 => unsafe { LLVMFloatTypeInContext(self.ctx) },
+            metal_ast::Primitives::F64 => unsafe { LLVMDoubleTypeInContext(self.ctx) },
+            metal_ast::Primitives::F128 => unsafe { LLVMFP128TypeInContext(self.ctx) },
+
+            metal_ast::Primitives::String => {
+                todo!()
+            }
+            metal_ast::Primitives::Void => unsafe { LLVMVoidTypeInContext(self.ctx) },
+        }
+    }
+
+    pub fn ty(&self, ty: &metal_ast::Ty) -> LLVMTypeRef {
+        match ty {
+            metal_ast::Ty::Primitive(primitive) => self.primitive(primitive),
+            _ => {
+                todo!()
+            }
+        }
+    }
+
+    pub fn function_definition(&self, ty: metal_ast::FnDefStmt, library: bool) -> LLVMValueRef {
+        let ret_ty = ty
+            .return_type
+            .unwrap_or(metal_ast::Ty::Primitive(metal_ast::Primitives::Void));
+        let return_type = self.ty(&ret_ty);
+
+        let mut params = Vec::with_capacity(ty.inputs.len());
+
+        for fn_input in ty.inputs {
+            params.push(self.ty(&fn_input.ty))
+        }
+
+        let linkage = if ty.public {
+            LLVMLinkage::LLVMExternalLinkage
+        } else {
+            LLVMLinkage::LLVMInternalLinkage
+        };
+
+        let fun_name: String = if library {
+            let mut name = "".to_string();
+            for module in self.module_name.split('.') {
+                let m1 = module.len().to_string() + module;
+                let m = m1.as_str();
+                name += m;
+            }
+            let n = name + ty.ident.inner.len().to_string().as_str() + ty.ident.inner;
+            n
```

Comment:
it's a bit confusing because I was running into a bunch of issues but basically, if current
module is a library, the name is
`module + function_name` (although every entry has a number before it corresponding to its length)

otherwise it's just `function_name`

Reply:
Aren't names user (compiler, us) -defined?

Reply:
Yes, but the name (module + function_name, so i.e. `3std2io5print`,) just removes some possibilities of name collisions across separate modules.

Reply:
Yeah, that's what name mangling is. I suggest extracting this logic into a separate function, possibly with a doc comment, to make code's intent clearer.